inferno-ml-server-types/src/Inferno/ML/Server/Types.hs

```diff
@@ -715,10 +717,19 @@ instance FromJSON IValue where
     Object o -> ITime <$> o .: "time"
     Array a
       | [x, y] <- Vector.toList a ->
-          fmap ITuple $
-            (,) <$> parseJSON x <*> parseJSON y
-      | Vector.null a -> pure IEmpty
-    _ -> fail "Expected one of: string, double, time, tuple, unit (empty array)"
+          (,) <$> parseJSON x <*> parseJSON y <&> \case
+            -- We don't want to confuse a two-element array of tuples with
+            -- a tuple itself. For example, `"[[10.0, {\"time\": 10}], [10.0, {\"time\": 10}]]"`
+            -- should parse as a two-element array of `(double, time)` tuples,
+            -- not as a `((double, time), (double, time))`. I can't think of
```

Comment:
Haha yeah I see the issue, and agree with your solution, but btw I think inferno allows n-tuples for any n != 1. As with my previous comment, I don't think Tach supports tuples, so they're only script-internal values and we wouldn't need to support them in the interface to Tach.